x-pack/metricbeat/module/openai/usage/_meta/fields.yml

```diff
@@ -0,0 +1,233 @@
+- name: usage
+  type: group
+  release: beta
+  description: >
+    OpenAI API usage metrics and statistics
+  fields:
+    - name: data
+      type: group
+      description: >
+        General usage data metrics
+      fields:
+        - name: organization_id
+          type: keyword
+          description: Organization identifier
+        - name: organization_name
+          type: keyword
+          description: Organization name
+        - name: aggregation_timestamp
+          type: date
+          description: Timestamp of data aggregation
+        - name: n_requests
+          type: long
+          description: Number of requests made
+        - name: operation
+          type: keyword
+          description: Operation type
+        - name: snapshot_id
+          type: keyword
+          description: Snapshot identifier
+        - name: n_context_tokens_total
+          type: long
+          description: Total number of context tokens used
+        - name: n_generated_tokens_total
+          type: long
+          description: Total number of generated tokens
+        - name: n_cached_context_tokens_total
+          type: long
+          description: Total number of cached context tokens
+        - name: email
+          type: keyword
+          description: User email
+        - name: api_key_id
+          type: keyword
+          description: API key identifier
+        - name: api_key_name
+          type: keyword
+          description: API key name
+        - name: api_key_redacted
+          type: keyword
+          description: Redacted API key
+        - name: api_key_type
+          type: keyword
+          description: Type of API key
+        - name: project_id
+          type: keyword
+          description: Project identifier
+        - name: project_name
+          type: keyword
+          description: Project name
+        - name: request_type
+          type: keyword
+          description: Type of request
+
+    - name: dalle
```

Comment:
@shmsr : We are having separate metrics for different type of models we have in OpenAI ?
Is that prudent ? Do we mean to say that all models would give different metrics ?
I see there are some metrics common across models like organization_id, api_key_type, project_id. Do we need separate metrics for each of these ?
Ideally these should be under one group and an additional field telling which model type these belong to.
WDYT ?

Reply:
> We are having separate metrics for different type of models (dall-e, whisper etc) we have in OpenAI ?
Is that prudent ? Do we mean to say that all models would give different metrics ?

Yes. They themselves have categorized like:

- Uni/ multi model to text (gpt 4o, gpt3, gpt2, etc.)
- Text to Image (dalle model)
- Text to Speech (tts)
- Speech to Text (whisper)
- Embeddings model

etc.

So, for GPT family, the metrics are like based on tokens. For dalle, it's the image dimension and number of images. For tts, it is number of characters synthesized (speech). For whisper, it is transcribed seconds.


See the attached image for example: <img width="1033" alt="Screenshot 2024-12-03 at 3 49 47 PM" src="https://github.com/user-attachments/assets/9ed8209e-467b-4be5-a81a-01db1665eae9">

So, as you can see very different type of metrics for each. So keeping the essential metrics under their respective subblocks like `data.*`, `dalle.*`, `whisper.*`, etc.

> I see there are some metrics common across models like organization_id, api_key_type, project_id. Do we need separate metrics for each of these ?

I agree with this. Now that we have a common struct across all 4 structs, we can move this to common; maybe directly under `usage.*`

```
type BaseData struct {
	OrganizationID   string  `json:"organization_id"`
	OrganizationName string  `json:"organization_name"`
	UserID           *string `json:"user_id"`
	ApiKeyID         *string `json:"api_key_id"`
	ApiKeyName       *string `json:"api_key_name"`
	ApiKeyRedacted   *string `json:"api_key_redacted"`
	ApiKeyType       *string `json:"api_key_type"`
	ProjectID        *string `json:"project_id"`
	ProjectName      *string `json:"project_name"`
}
```

We can move the `model_id` too in `usage.*`. It is just that by default, OpenAI is sending the model-name  for GPT models in `snapshot_id` and for rest TTS, whisper, etc. it comes in mode_id. So we can move that too directly under `usage.*`.

> WDYT ?

So, agree with moving the common metrics under `usage.*` instead of having them under in respective broad categories of models. And keep the rest same.

Reply:
Let me know if you have more suggestions.

Reply:
So for the models which will not be running, the metrics will not be fetched or will the values be 0?

Reply:
Thanks, we can go ahead with the plan discussed above. 
The only concern would be that we would have to add metrics to the beats code for any new model that openai would support in future. 
Is my assumption right

Reply:
Yes, this is a problem in case they add a different category of model. We have to fix it in beats.